### PR TITLE
Pickup notification

### PR DESCRIPTION
**Pull request by gvszg**

### CODE REVIEW

File: app/services/gcm_notify_service.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    DeviceRegistration.select(:id, :registration_id).find_in_batches do |ids|
+      registration_ids = ids.map(&:registration_id)        
+      self.gcm.send_notification(registration_ids, options)
+    end
+  end
+
+  def send_pickup_notification(order)
+    device_id = DeviceRegistration.find_by(user_id: order.user_id).registration_id
+    registration_id = device_id.split
```
Comment (StevenKo) [inline app/services/gcm_notify_service.rb]: [device_id]

File: app/controllers/api/v1/orders_controller.rb [hunk 2]
```diff
@@ -8,8 +8,20 @@ def create
...
+      device_of_order = DeviceRegistration.find_or_initialize_by(registration_id: order.registration_id)
+      if device_of_order.new_record?
+        device_of_order.user = order.user
+        device_of_order.save!
+        logger.info "The new device has been saved!"
+      elsif device_of_order.user.nil?
+        device_of_order.update_attributes!(user_id: order.user_id)
+        logger.info "The user id: #{device_of_order.user_id} has been updated to the device!"
+      end
+
```
Comment (StevenKo) [inline app/controllers/api/v1/orders_controller.rb:19]: device_of_order.user = order.user
device_of_order.save!

File: app/services/gcm_notify_service.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  def send_item_event_notification(notification)
+    options = generate_options_for_item(notification)
+    DeviceRegistration.select(:id, :registration_id).find_in_batches do |ids|
+      registration_ids = ids.map(&:registration_id)
+      self.gcm.send_notification(registration_ids, options)
+    end
+  end
+
+  def send_pickup_notification(order)
+    device_id = DeviceRegistration.find_by(registration_id: order.registration_id).registration_id
```
Comment (StevenKo) [inline app/services/gcm_notify_service.rb]: order.device

File: app/controllers/api/v1/orders_controller.rb [hunk 4]
```diff
@@ -8,6 +8,11 @@ def create
       order.items_price = params[:items_price]
       order.ship_fee = params[:ship_fee]
       order.total = params[:total]
+      # Save user_id to device_registrations table
+      device_of_order = DeviceRegistration.find_or_initialize_by(registration_id: params[:registration_id])
+      device_of_order.user = order.user
+      device_of_order.save!
+      order.device_registration = device_of_order
```
Comment (StevenKo) [inline app/controllers/api/v1/orders_controller.rb]: unless params[:registration_id].nil?
  code
end

File: app/services/gcm_notify_service.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      registration_ids = ids.map(&:registration_id)
+      self.gcm.send_notification(registration_ids, options)
+    end
+  end
+
+  def send_pickup_notification(order)
+    device_id = order.device_registration.registration_id
+    registration_id = [device_id]
+    options = generate_options_for_pickup(order)
+    self.gcm.send(registration_id, options)
```
Comment (StevenKo) [inline app/services/gcm_notify_service.rb]: self.gcm is equal to @gcm ?

File: app/services/gcm_notify_service.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        content_title: notification.content_title,
+        content_text: notification.content_text,
+        content_pic: notification.send_content_pic,
+        item_id: notification.item_id,
+        item_name: notification.item.name,
+        item_price: notification.item.price
+        },
+      collapse_key: "updated_score"
+    }
+    return result_options
```
Comment (varbark) [inline app/services/gcm_notify_service.rb]: 不需要return result_options 
因為在ruby最後一行code 會自動被return

**Pull request merged**